pkg/destroy/aws/shared.go

```diff
@@ -0,0 +1,233 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+func (o *ClusterUninstaller) removeSharedTags(
+	ctx context.Context,
+	session *session.Session,
+	tagClients []*resourcegroupstaggingapi.ResourceGroupsTaggingAPI,
+	tracker *errorTracker,
+) error {
+	for _, filter := range o.Filters {
+		for key, value := range filter {
+			if strings.HasPrefix(key, "kubernetes.io/cluster/") {
+				if value == "owned" {
+					if err := o.removeSharedTag(ctx, session, tagClients, key, tracker); err != nil {
+						return err
+					}
+				} else {
+					o.Logger.Warnf("Ignoring non-owned cluster key %s: %s for shared-tag removal", key, value)
+				}
+			}
+		}
+	}
+
+	return nil
```

Comment:
Yes, this is a bit of code that always takes me too long to understand every time that I look at it. I don't know that we want to rely on the `clusterID` being set correctly, although I can't think of a good reason why it wouldn't be.

Maybe it would be sufficient to have a function that gets the cluster tag keys from the filter.
```
func clusterOwnedKeys(filters []Filter) []string {
	var keys[] string
	for _, filter := range filters {
		for key, value := range filter {
			if !strings.HasPrefix(key, "kubernetes.io/cluster/") {
				continue
			}
			if value != "owned" {
				continue
			}
			keys = append(keys, key)
		}
	}
	return keys
}
```
And then the `removeSharedTags` function could be the following.
```
func removeSharedTags(ctx context.Context, tagClients []*resourcegroupstaggingapi.ResourceGroupsTaggingAPI, filters []Filter, logger logrus.FieldLogger) error {
	for _, key := range clusterOwnedKeys(filter) {
		if err := removeSharedTag(ctx, tagClients, key, logger); err != nil {
			return err
		}
	}
	return nil
}
```